File: 백준/Gold/16973. 직사각형 탈출/직사각형 탈출.py
import sys
input = sys.stdin.readline
from collections import deque

def check(x, y):
    for x_w, y_w in walls:
        if x <= x_w < x + h and y <= y_w < y + w:
            return False
    return True

def bfs(x, y):
    q = deque()
    q.append((x, y, 0))
    visited = [[False] * m for _ in range(n)]

    while q:
        now_x, now_y, cnt = q.popleft()
        visited[now_x][now_y] = True

        if now_x == fr-1 and now_y == fc-1:
            return cnt
        
        for i in range(4):
            next_x = now_x + dir[i][0]
            next_y = now_y + dir[i][1]

            if 0 <= next_x < n and 0 <= next_y < m and 0 <= next_x + h -1 < n and 0 <= next_y + w - 1 < m:
                if not visited[next_x][next_y]:
                    if check(next_x, next_y):
                        visited[next_x][next_y] = True
                        q.append((next_x, next_y, cnt+1))
    return -1

n, m = map(int,input().split())
dir = [[-1, 0], [1, 0], [0, -1], [0, 1]]
graph = []
for _ in range(n):
    graph.append(list(map(int, input().split())))

h, w, sr, sc, fr, fc = map(int, input().split())

walls = []
for i in range(n):
    for j in range(m):
        if graph[i][j] == 1:
            walls.append((i,j))

print(bfs(sr-1, sc-1))

File: 백준/Gold/16973. 직사각형 탈출/직사각형 탈출.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

class Square{
    int x;
    int y;
    int cnt;

    public Square(int x, int y, int cnt) {
        this.x = x;
        this.y = y;
        this.cnt = cnt;
    }
}

public class Main {
    static int n, m;
    static int h,w,sr,sc,fr,fc;
    static int[][] dir = {{-1,0}, {1, 0}, {0, -1}, {0, 1}};
    static int[][] map;

    static boolean[][] visited;
    static List<int[]> wall;
    static int[] now;
    public static boolean check(int x, int y) {
        for (int i = 0; i < wall.size(); i++) {
            now = wall.get(i);
            if (x <= now[0] && now[0] <= x + h - 1 && y <= now[1] && now[1] <= y + w - 1) {
                return false;
            }
        }
        return true;
    }

    public static int bfs() {
        Queue<Square> queue = new LinkedList<>();
        queue.add(new Square(sr, sc, 0));
        visited[sr][sc] = true;

        while (!queue.isEmpty()) {
            Square square = queue.poll();


            if (square.x == fr && square.y == fc) {
                return square.cnt;
            }

            for (int i = 0; i < 4; i++) {
                int nx = square.x + dir[i][0];
                int ny = square.y + dir[i][1];

                if (nx < 0 || ny < 0 || nx >= n || ny >= m || visited[nx][ny])
                    continue;
                if (nx + h - 1 >= n || ny + w - 1 >= m)
                    continue;

                if (check(nx, ny)) {
                    visited[nx][ny] = true;
                    queue.add(new Square(nx, ny, square.cnt + 1));
                }
            }
        }
        return -1;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        map = new int[n][m];
        wall = new ArrayList<>();
        visited = new boolean[n][m];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            for (int j = 0; j < m; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());

                if (map[i][j] == 1) {
                    wall.add(new int[] {i, j});
                }
            }
        }

        st = new StringTokenizer(br.readLine());
        h = Integer.parseInt(st.nextToken());
        w = Integer.parseInt(st.nextToken());
        sr = Integer.parseInt(st.nextToken()) - 1;
        sc = Integer.parseInt(st.nextToken()) - 1;
        fr = Integer.parseInt(st.nextToken()) - 1;
        fc = Integer.parseInt(st.nextToken()) - 1;

        System.out.println(bfs());
    }
}